src/interlocking_component/route_controller.py

```diff
@@ -609,8 +609,10 @@ def maybe_put_reservations_as_first(self, train: Train, route: List[Node]):
         for i, edge in enumerate(edge_route):
             if not edge.track.is_reservation_track:
                 continue
-            if edge.track.reservations[0][0].edge == edge:
-                # The train this is reserved for is already on the swgment, that may be swaped.
+            first_reserving_train = edge.track.reservations[0][0]
```

Comment:
Does this part check for other trains in the block section too?

Reply:
Not exactly. It checks if the train is on the `ReservationTrack` or close to it. Close in this context means, that this `ReservationTrack` is the next `ReservationTrack`, the train has an reservation for. Each block section apart from those crossing switches has exactly one `ReservationTrack`. If the train is not on it's previous `ReservationTrack`, the check for this train, if it has the neccassary reservations, has checked the `ReservationTrack` that may be swaped and because of that has allowed the train to continue driving. If the reservation is then swaped away from first position, the train may drive toward a `ReservationTrack`, that is not reserved for it in first position. This will lead to a situation, where the first reserving train is stuck behind the second reserving train.
To prevent that we need to make sure, that prior to swaping reservations the first reserving train is still on or driving towards another `ReservationTrack`. If the reservations get swapped then, this train will not be allowed to drive towards the `ReservationTrack` where reservations were swapped, because it does not have the reservation in first position there.
It is not importent, if a train is on the same blocksection after the `ReservationTrack`, because it can always just leave and crashes are prevented by the use of the interlocking.